### PR TITLE
Dockerize services

### PR DESCRIPTION
**Pull request by adaszko**
A tenative version.  Everything builds, but hasn’t been tested working together.

To do before merging:
 * [ ] Copy over the GCR_JSON_KEY secret from the https://github.com/radicle-dev/radicle-client-services/

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+version: "3.7"
+services:
+  radicle-node:
+    image: gcr.io/radicle-services/radicle-node:${RADICLE_IMAGE_TAG:-latest}
+    build:
+      dockerfile: ./radicle-node/Dockerfile
+      context: .
+    environment:
+      RUST_LOG: hyper=warn,debug
```
Comment (cloudhead) [inline docker-compose.yml:9]: We can just set this to `debug` as there's no hyper in the node.

Comment (adaszko) [reply docker-compose.yml:9]: Gotcha 👍

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          memory: 6gb
+  caddy:
+    image: caddy:2.4.5
+    entrypoint:
+      - sh
+      - -euc
+      - |
+        cat <<EOF >/etc/caddy/Caddyfile
+        {$RADICLE_DOMAIN} {
+            reverse_proxy git-httpd:8778
```
Comment (cloudhead) [inline docker-compose.yml:43]: Where is `git-httpd` coming from? Should it be `radicle-httpd`?

File: radicle-httpd/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+FROM rust:1.65.0-slim@sha256:27349440d16ac15f52e8cae6181999a1b2fc05272d612a358d0e53078fcef88e as build
+
+RUN apt-get update && apt-get install -y pkg-config libssl-dev git cmake
+
+COPY . /app
+WORKDIR /app/radicle-httpd
+
+RUN set -eux; \
+    cargo install --locked --path .; \
+    objcopy --compress-debug-sections /usr/local/cargo/bin/radicle-httpd /usr/local/cargo/bin/radicle-httpd.compressed
```
Comment (cloudhead) [inline radicle-httpd/Dockerfile:11]: Can we use multiple `RUN` lines here instead of `\`?

File: radicle-node/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Build
+FROM rust:1.65.0-slim@sha256:27349440d16ac15f52e8cae6181999a1b2fc05272d612a358d0e53078fcef88e as build
+
+RUN apt-get update && apt-get install -y pkg-config libssl-dev git cmake
```
Comment (cloudhead) [inline radicle-node/Dockerfile:4]: It's worth checking if it builds without `cmake` and `pkg-config` now

File: docker-compose.yml (hunk 2 above)
Comment (adaszko) [reply docker-compose.yml:43]: Yep, that would be a typo!

File: radicle-httpd/Dockerfile (hunk 3 above)
Comment (adaszko) [reply radicle-httpd/Dockerfile:11]: Sure thing

File: radicle-node/Dockerfile (hunk 4 above)
Comment (adaszko) [reply radicle-node/Dockerfile]: This should be resolved now

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  build-and-push-images:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Login to the container registry
+        uses: docker/login-action@v1
+        with:
+          registry: gcr.io
+          username: _json_key
```
Comment (cloudhead) [inline .github/workflows/deploy.yml:18]: This looks really weird. Is it correct?

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: radicle-node/Dockerfile
+          push: true
+          tags: gcr.io/radicle-services/radicle-node:latest,gcr.io/radicle-services/radicle-node:${{ github.sha }}
+          cache-from: type=registry,ref=gcr.io/radicle-services/radicle-node:latest
+          cache-to: type=inline
+      - name: Build and push radicle-httpd
+        id: http_api
```
Comment (cloudhead) [inline .github/workflows/deploy.yml:33]: Can we change this to `radicle-httpd` as well?

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (adaszko) [reply .github/workflows/deploy.yml:18]: Corrent, it requires copying over the `GCR_JSON_KEY` from the https://github.com/radicle-dev/radicle-client-services/ repo before we first use this job to deploy though

File: .github/workflows/deploy.yml (hunk 6 above)
Comment (adaszko) [reply .github/workflows/deploy.yml]: Done ✅

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (cloudhead) [reply .github/workflows/deploy.yml:18]: Hmf ok.

**Comment by cloudhead** [comment]
Is there anything else that needs to be done here before we can merge?

**Comment by adaszko** [comment]
> Is there anything else that needs to be done here before we can merge?

It’s safe to merge in the sense it won’t break anything, so I think we can merge.

**Pull request merged**